rust/kernel/src/sync.rs

```diff
@@ -0,0 +1,125 @@
+// SPDX-License-Identifier: GPL-2.0
+
+use core::cell::UnsafeCell;
+use core::ops::{Deref, DerefMut};
+use core::fmt;
+
+use crate::bindings;
+
+#[cfg(CONFIG_DEBUG_SPINLOCK)]
+compile_error!("CONFIG_DEBUG_SPINLOCK not yet implemented");
+
+#[cfg(CONFIG_DEBUG_MUTEX)]
+compile_error!("CONFIG_DEBUG_MUTEX not yet implemented");
+
+pub struct Mutex<T: Send> {
+    lock: bindings::mutex,
+    data: UnsafeCell<T>,
+}
+
+impl<T: Send> Mutex<T> {
+    /// Create a new Mutex
+    pub fn new(data: T, name: &'static str) -> Self {
+
+        let lock = bindings::mutex::default();
+        // TODO: debug foo
+
+        Self {
+            data: UnsafeCell::new(data),
+            lock
+        }
+    }
+
+    /// acquire a lock on the mutex
+    pub fn lock<'a>(&'a self) -> MutexGuard<'a, T> {
+        unsafe {
+            bindings::mutex_lock(&self.lock as *const bindings::mutex as *mut bindings::mutex);
+        }
+        MutexGuard { inner: &self }
+    }
+
+    /// try to acquire the lock, returns none on failure
+    pub fn try_lock<'a>(&'a self) -> Option<MutexGuard<'a, T>> {
+        let ret = unsafe { bindings::mutex_trylock(&self.lock as *const bindings::mutex as *mut bindings::mutex) };
+        if ret == 1 {
+            Some(MutexGuard { inner: &self })
+        } else {
+            None
+        }
+    }
+
+    /// test if the mutex is locked
+    pub fn is_locked(&self) -> bool {
+        unsafe {
+            bindings::mutex_is_locked(&self.lock as *const bindings::mutex as *mut bindings::mutex)
+        }
+    }
+
+    fn unlock(&self) {
+        unsafe {
+            bindings::mutex_unlock(&self.lock as *const bindings::mutex as *mut bindings::mutex);
+        }
+    }
+}
+
+unsafe impl<T: Send> Send for Mutex<T> { }
+unsafe impl<T: Send> Sync for Mutex<T> { }
+
+impl<T: Send + fmt::Debug> fmt::Debug for Mutex<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        // kindly borrowed from std::sync::Mutex
+        match self.try_lock() {
+            Some(guard) => f.debug_struct("Mutex").field("data", &guard).finish(),
+            None => {
+                struct LockedPlaceholder;
+                impl fmt::Debug for LockedPlaceholder {
+                    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+                        f.write_str("<locked>")
+                    }
+                }
+
+                f.debug_struct("Mutex").field("data", &LockedPlaceholder).finish()
+            },
+        }
+    }
+}
+
+pub struct MutexGuard<'a, T: Send> {
+    inner: &'a Mutex<T>,
+}
+
+
+unsafe impl<'a, T: Send> Send for MutexGuard<'a, T> { }
+unsafe impl<'a, T: Send> Sync for MutexGuard<'a, T> { }
```

Comment:
I don't think these are right, these would let you have multiple concurrent copies in multiple threads. Here's what the stdlib `MutexGuard` has:

![image](https://user-images.githubusercontent.com/772/100556733-5e9dea00-3272-11eb-974d-f648ac0de8e3.png)

Reply:
I think I got it now.